File: 0123-best-time-to-buy-and-sell-stock-iii/0123-best-time-to-buy-and-sell-stock-iii.ts
function maxProfit(prices: number[]): number {
  let ans = 0
  const fw: number[] = new Array(prices.length).fill(0)
  const bw: number[] = new Array(prices.length).fill(0)
  let lMin = prices[0]
  let rMax = prices.at(-1)
  for (let i = 1; i < prices.length; i++) {
    fw[i] = Math.max(fw[i - 1], prices[i] - lMin)
    lMin = Math.min(lMin, prices[i])
  }
  for (let i = prices.length - 2; i >= 0; i--) {
    bw[i] = Math.max(bw[i + 1], rMax - prices[i])
    rMax = Math.max(rMax, prices[i])
  }
  for (let i = 0; i < prices.length; i++) {
    ans = Math.max(ans, fw[i] + bw[i])
  }
  return ans
}

File: 0123-best-time-to-buy-and-sell-stock-iii/0123-best-time-to-buy-and-sell-stock-iii.go
func maxProfit(prices []int) int {
  ans, fw, bw := 0, make([]int, len(prices)), make([]int, len(prices))
  fw[0], bw[len(bw) - 1] = 0, 0
  lMin, rMax := prices[0], prices[len(prices) - 1]
  for i := 1; i < len(prices); i++ {
    fw[i] = max(fw[i - 1], prices[i] - lMin)
    lMin = min(lMin, prices[i])
  }
  for i := len(prices) - 2; i >= 0; i-- {
    bw[i] = max(bw[i + 1], rMax - prices[i])
    rMax = max(rMax, prices[i])
  }
  for i := 0; i < len(prices); i++ {
    ans = max(ans, fw[i] + bw[i])
  }
  return ans
}